### DISCUSSION TITLE
How to support Ipv6 wildcard by using with_host config.

### DISCUSSION THREAD

**Discussion by jackymao** [trillium-rs/trillium#95] [Q&A] [answered]
Hi,

First of all, thank you very much for making this great framework/platform. After reading the wonderful guide on trillium.rs and the examples code, it feels amazing that it is really concise and also powerful, and extendable with the State, Conn, Handler concept.  

I have tried trillium for some days and am quite happy with its straightforward on things likes router, cookie/session, Askama template support, and the proxy is a great experience. I have also tried to work with OAuth2 and it works very well. Also to use with Casbin and get RBAC (Role-based access control) works fine. All these are so good that it just works.  

As said if you like trillium then using it. I made a website and try to deploy it to fly.io. There is an example app from Fly.io using warp and I am trying to make it works with trillium. It is required to bind to all address by ([0,0,0,0,0,0,0,0], 8080) as [ref: https://github.com/fly-apps/hello-rust](https://github.com/fly-apps/hello-rust). I meet problem here that I try as below:  
```rust
trillium_smol::config()
    .with_port(8080)
    .with_host("0.0.0.0")  // no matter 4 zeros or 6 zeros or 8 zeros
    .run(handler)
```
this is ok on local dev but can't up on fly.io. I don't know if it is an issue about the IPv6 wildcard.  

Would you please help check this and see what I can do on this. Many thanks.

**Comment by jbr** [comment, marked as answer]
Thanks for using trillium! I believe that `0,0,0,0,0,0` is the same as [std::net::Ipv6Addr::UNSPECIFIED](https://doc.rust-lang.org/std/net/struct.Ipv6Addr.html#associatedconstant.UNSPECIFIED) which is parsed as `"::"` — so if you set host to `"::"`, I believe that it should correctly bind to `0,0,0,0,0,0`

If that does work, please let me know and I'll add a note to the documentation, because it's not at all obvious

**Comment by jackymao** [comment]
Thanks for your prompt reply. It is a surprise when seeing the reply in less than 5 minutes after I post this discussion. Amazing. 

I have tried `with_host("::")` and it works.  

And finally I found the problem is on deploying process that is regarding health check, not coding problem. Actually both `with_host("0.0.0.0")` and `with_host("::")` are working, while one is for IPv4 and the other is for both IPv4 and IPv6 in my case at fly.io.  

Thanks again for your help. And I will report here if any other discussion on trillium.

**Reply by jackymao** [reply]
-------
update (2021-08-03):
Sorry but have to report an exception that on my another local Win7 dev computer (can't visit IPv6 website):  
- `with_host("0.0.0.0")`: I can visit the site locally with `localhost:8080` and `127.0.0.1:8080`  
- `with_host("::")`: `localhost:8080` is OK, but `127.0.0.1:8080` not working.  

So maybe fly.io have some special setting on the IPv6 & IPv4 service.  
Is there a way to have the  `with_host("0.0.0.0")` and `with_host("::")`  double setting?

**Reply by jbr** [reply]
Currently, the way to support that would be to use the HOST environment variable to customize the deployments. You should be able to use the default configuration (`trillium_smol::run(…)`) with `HOST="::"` or with no setting at all to bind to localhost in development. It looks like https://fly.io/docs/reference/configuration/#the-env-variables-section would be the way to provide `HOST="::"` for fly.io

**Reply by jackymao** [reply]
Appreciate that you spend time on this and even check the fly.io document. 

Currently `with_host("::")` on fly.io works very well for both IPv4 and IPv6. And now `with_host("0.0.0.0")` can up and running also and is fairly good for me. On local development, `with_host("0.0.0.0")` is also satisfied my requirement. 

Thank you so much for your kindly help!

Have a nice day.
